timemachine/fe/free_energy.py

```diff
@@ -951,30 +1022,35 @@ def log_q_fn(replica_idx: ReplicaIdx, state_idx: StateIdx) -> float:
     if barostat is not None:
         barostat.set_interval(equil_barostat_interval)
 
-    def get_equilibrated_xvb(xvb: CoordsVelBox, params: NDArray) -> CoordsVelBox:
-        if md_params.n_eq_steps == 0:
-            return xvb
-        # Ensure initial mover state is consistent across replicas
-        for mover in context.get_movers():
-            mover.set_step(0)
-
-        context.set_x_t(xvb.coords)
-        context.set_v_t(xvb.velocities)
-        context.set_box(xvb.box)
+    def get_equilibrated_xvb(xvb: CoordsVelBox, state_idx: StateIdx) -> CoordsVelBox:
+        # Set up parameters of the water sampler movers
+        if md_params.water_sampling_params is not None:
+            for mover in context.get_movers():
+                if isinstance(mover, WATER_SAMPLER_MOVERS):
+                    assert water_params_by_state is not None
+                    mover.set_params(water_params_by_state[state_idx])
+        if md_params.n_eq_steps > 0:
+            # Set the movers to 0 to ensure they all equilibrate the same way
+            # Ensure initial mover state is consistent across replicas
+            for mover in context.get_movers():
+                mover.set_step(0)
 
-        assert len(context.get_potentials()) == 1
-        context.get_potentials()[0].set_params(params)
+            params = params_by_state[state_idx]
+            bound_potentials[0].set_params(params)
 
-        xs, boxes = context.multiple_steps(md_params.n_eq_steps, store_x_interval=0)
-        x0 = xs[0]
-        v0 = context.get_v_t()
-        box0 = boxes[0]
+            context.set_x_t(xvb.coords)
+            context.set_v_t(xvb.velocities)
+            context.set_box(xvb.box)
 
-        return CoordsVelBox(x0, v0, box0)
+            xs, boxes = context.multiple_steps(md_params.n_eq_steps, store_x_interval=0)
+            x0 = xs[0]
+            v0 = context.get_v_t()
+            box0 = boxes[0]
```

Comment:
Is xs[0] == xs[-1] in this case? Just wondering why v0 is get_v_t() which would be the last frame right?

Reply:
Yes, the length of `xs[0]` should be one, will add an assertion to be sure. Its only that we don't return the velocities from `multiple_steps`.

Reply:
A bit annoying/confusing since the internal Context call disallows `store_x_interval=0`:

```
std::array<std::vector<double>, 2> Context::multiple_steps(const int n_steps, int store_x_interval) {
    if (store_x_interval <= 0) {
        throw std::runtime_error("store_x_interval <= 0");
    }
```

But the internal wrapper code converts it to `n_steps`:

```        
.def(
    "multiple_steps",
    [](Context &ctxt, const int n_steps, int store_x_interval) -> py::tuple {
        // (ytz): I hate C++
        int x_interval = (store_x_interval <= 0) ? n_steps : store_x_interval;
        std::array<std::vector<double>, 2> result = ctxt.multiple_steps(n_steps, x
```

So like @badisa  - I think this is just returning the last frame.

Reply:
Yup and we do have this logic tested: https://github.com/proteneer/timemachine/blob/master/tests/test_md.py#L65-L70